src/darwin/Framework/CHIP/MTROTAProviderDelegateBridge.mm

```diff
@@ -608,81 +608,85 @@ bool GetPeerNodeInfo(CommandHandler * commandHandler, const ConcreteCommandPath
-                        handle.Release();
-                        gOtaSender.ResetState();
-                        return;
-                    }
-                    auto targetNodeId
-                        = handler->GetExchangeContext()->GetSessionHandle()->AsSecureSession()->GetLocalScopedNodeId();
+    auto completionHandler
+        = ^(MTROTASoftwareUpdateProviderClusterQueryImageResponseParams * _Nullable data, NSError * _Nullable error) {
+              [controller
+                  asyncDispatchToMatterQueue:^() {
+                      assertChipStackLockedByCurrentThread();
 
-                    char uriBuffer[kMaxBDXURILen];
-                    MutableCharSpan uri(uriBuffer);
-                    err = bdx::MakeURI(targetNodeId.GetNodeId(), AsCharSpan(data.imageURI), uri);
-                    if (CHIP_NO_ERROR != err) {
-                        LogErrorOnFailure(err);
-                        handler->AddStatus(cachedCommandPath, Protocols::InteractionModel::Status::Failure);
-                        handle.Release();
-                        gOtaSender.ResetState();
-                        return;
-                    }
+                      CommandHandler * handler = EnsureValidState(handle, cachedCommandPath, "QueryImage", data, error);
+                      VerifyOrReturn(handler != nullptr);
 
-                    response.imageURI.SetValue(uri);
-                    handler->AddResponse(cachedCommandPath, response);
-                    handle.Release();
-                    return;
-                }
-                if (!hasUpdate) {
-                    // Send whatever error response our delegate decided on.
-                    handler->AddResponse(cachedCommandPath, response);
-                } else {
-                    // We must have isBDXProtocolSupported false.  Send the corresponding error status.
-                    Commands::QueryImageResponse::Type protocolNotSupportedResponse;
-                    protocolNotSupportedResponse.status
-                        = static_cast<OTAQueryStatus>(MTROTASoftwareUpdateProviderOTAQueryStatusDownloadProtocolNotSupported);
-                    handler->AddResponse(cachedCommandPath, protocolNotSupportedResponse);
-                }
-                handle.Release();
-                gOtaSender.ResetState();
-            }
+                      ChipLogDetail(Controller, "QueryImage: application responded with: %s",
+                          [[data description] cStringUsingEncoding:NSUTF8StringEncoding]);
 
-                          errorHandler:^(NSError *) {
-                              // Not much we can do here
-                          }];
-    };
+                      Commands::QueryImageResponse::Type delegateResponse;
+                      ConvertFromQueryImageResponseParms(data, delegateResponse);
+
+                      // If BDX protocol is not supported, send a response with status - Protocol Not supported and reset state.
```

Comment:
if the requestor gets protocol not supported, the spec says it may re-query another provider in the list. Can we read that as it cannot re-query the same provider? This is the reason i put the protocol not supported at the top of the list. if protocol is not supported for this query image, return error reset state. Then we have the update available or not case. Thats how I read it but if my understanding is not correct then we can check for available first and follow through with the logic you suggested.

If the OTA Requestor is still attempting to discover an OTA Update, it MAY choose to attempt a [QueryImage command](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/7717261055697364e460d846ae65b163e359181b/src/service_device_management/OTASoftwareUpdate.adoc#QueryImageCommand) on a different OTA Provider in its OTA Provider List, in which case the timing for the query SHALL match the query timing constraints expressed in the previous paragraphs of this section. Otherwise, it SHALL continue to query the same OTA Provider, again following the query timing constraints previously expressed.

Reply:
> Can we read that as it cannot re-query the same provider?

No, I don't think we can.  Especially since the next sentence says:

>  Otherwise, it SHALL continue to query the same OTA Provider, again following the query timing constraints previously expressed.

Here is the real question: is a requestor that prefers HTTPS to BDX but supports both allowed to query with just HTTPS and then if that fails with NotSupported re-query with BDX (which is required to be supported)?  That's the situation I am worrying about here.  If requestors can do that, we want to prioritize "no image" over "we don't support that protocol" so they don't do the extra request.

Reply:
I would prefer protocol not supported first. since i do not support the protocol you asked for, therefore I do not have an image I can transfer to you. I can double check with Carol/Phil if they have any thoughts on this.

Reply:
Rethinking this and prioritizing image not available first because it makes more sense.

Reply:
Checked with Phil. He agrees to check for protocol not supported first. Please let me know if you feel strongly about this and we can discuss :)

Reply:
I'd like to understand the reasoning.

Reply:
added a comment